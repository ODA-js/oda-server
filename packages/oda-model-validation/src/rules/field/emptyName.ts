import { IValidationResult } from '../../../interfaces';
import { IFieldContext } from '../../interfaces';
import { Rule } from '../../rules';

export default class implements Rule<IFieldContext> {
  public name = 'field-empty-name';
  public description = 'field must be named';
  public validate(context: IFieldContext): IValidationResult[] {
    const result: IValidationResult[] = [];
    if (!context.field.name) {
      result.push({
        message: this.description,
        result: 'error',
      });
    }
    return result;
  }
}
