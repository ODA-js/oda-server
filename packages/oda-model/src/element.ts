import { IValidate, IValidator, IValidationResultInput } from './validation';
import { merge } from 'lodash';
import { MetaModelType, Nullable, assignValue } from './types';

export const Internal = Symbol.for('internal');
export const MetaData = Symbol.for('metadata');

export interface ElementMetaInfo {}

export interface ElementInternal<T extends ElementMetaInfo> {
  [MetaData]: T;
}

export interface ElementInput<T extends ElementMetaInfo> {
  metadata?: Partial<T>;
}

export interface ElementOutput<T extends ElementMetaInfo> {
  metadata: T;
}

export interface IUpdatableBase {
  /**
   * update item with data
   * @param payload the update payload
   */
  updateWith(payload: any): void;
  /**
   * return copy of object item that is suitable for creating new one
   */
  toObject(): any;
}

/**
 * updatable items
 */
export interface IUpdatable<
  T extends ElementMetaInfo,
  I extends ElementInput<T>,
  O extends ElementOutput<T>
> extends IUpdatableBase {
  /**
   * update item with data
   * @param payload the update payload
   */
  updateWith(payload: Nullable<I>): void;
  /**
   * return copy of object item that is suitable for creating new one
   */
  toObject(): O;
}

/**
 * Meta interface for updating items with metadata
 */
export interface IMeta<
  M extends ElementMetaInfo,
  I extends ElementInput<M>,
  O extends ElementOutput<M>
> extends IUpdatable<M, I, O>, IValidate {
  /**
   * the kind of current item
   */
  readonly modelType: MetaModelType;
  /**
   * meta information that is outside of model notations and can be customized as well
   */
  readonly metadata: Readonly<M>;
}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Element<
  M extends ElementMetaInfo,
  I extends ElementInput<M>,
  S extends ElementInternal<M>,
  O extends ElementOutput<M>
> implements IMeta<M, I, O> {
  public get modelType(): MetaModelType {
    return 'element';
  }
  public [Internal]: S;
  /**
   * metadata is not immutable
   */
  public get metadata(): M {
    return this[Internal][MetaData];
  }

  public validate(validator: IValidator): IValidationResultInput[] {
    return validator.check(this);
  }

  constructor(init: ElementInput<M>) {
    this[Internal] = {} as S;
    this.updateWith(merge({}, defaultInput, init));
  }
  /**
   * patches metadata for entity
   * @param input metadata patch object
   */
  public updateWith(input: Nullable<ElementInput<M>>) {
    assignValue<Element<M, I, S, O>, ElementInput<M>, ElementInternal<M>>({
      src: this,
      input,
      field: 'metadata',
      effect: (_, value) =>
        (this[Internal][MetaData] = merge({}, this[Internal][MetaData], value)),
      setDefault: _ =>
        (this[Internal][MetaData] = merge({}, defaultMetaInfo) as M),
    });
  }
  /**
   * make item usable within another constructor
   * it is suitable for cloning items
   */
  public toObject(): O {
    return merge({}, { metadata: this.metadata } as O);
  }
}
