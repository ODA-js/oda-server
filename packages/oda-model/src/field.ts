import {
  SimpleField,
  SimpleFieldInput,
  ISimpleField,
  SimpleFieldOutput,
} from './simplefield';
import {
  EntityField,
  EntityFieldInput,
  IEntityField,
  EntityFieldOutput,
} from './entityfield';
import {
  RelationField,
  RelationFieldInput,
  IRelationField,
  RelationFieldOutput,
} from './relationfield';

export type Field = SimpleField | EntityField | RelationField;
export type FieldInput =
  | SimpleFieldInput
  | EntityFieldInput
  | RelationFieldInput;

export type FieldOutput =
  | SimpleFieldOutput
  | EntityFieldOutput
  | RelationFieldOutput;

export type IField = ISimpleField | IEntityField | IRelationField;

export function isEntityFieldInput(
  input: FieldInput,
): input is EntityFieldInput {
  const ent = (input as unknown) as EntityField;
  return ent.type && ent.type.type === 'entity';
}

export function isRelationFieldInput(
  input: FieldInput,
): input is RelationFieldInput {
  return !!(input as RelationFieldInput).relation;
}
export function isSimpleInput(input: FieldInput): input is SimpleFieldInput {
  return !(isEntityFieldInput(input) || isRelationFieldInput(input));
}
