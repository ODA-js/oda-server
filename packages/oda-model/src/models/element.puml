@startuml

interface IMeta
interface IUpdatable
interface IValidate

IUpdatable <|-- IMeta
IValidate <|-- IMeta

class Element

IMeta <|-- Element


@enduml
