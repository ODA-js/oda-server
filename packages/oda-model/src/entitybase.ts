import { ElementMetaInfo, Internal } from './element';
import {
  IModelBase,
  ModelBaseInput,
  ModelBaseInternal,
  ModelBase,
  ModelBaseOutput,
} from './modelbase';
import {
  AsHash,
  MetaModelType,
  Nullable,
  assignValue,
  HashToArray,
  NamedArray,
  MapToArray,
} from './types';
import { OperationInput, IOperation, Operation } from './operation';
import {
  FieldInput,
  IField,
  isSimpleInput,
  Field,
  isRelationFieldInput,
} from './field';
import { merge } from 'lodash';
import * as inflected from 'inflected';
import { SimpleField, SimpleFieldInput } from './simplefield';
import { DEFAULT_ID_FIELD } from './definitions';
import { RelationField, RelationFieldInput } from './relationfield';
import { EntityField, EntityFieldInput } from './entityfield';
import capitalize from './lib/capitalize';

export interface IEntityBase<
  T extends EntityBaseMetaInfo<P>,
  P extends EntityBasePersistence,
  I extends EntityBaseInput<T, P>,
  O extends EntityBaseOutput<T, P>
> extends IModelBase<T, I, O> {
  readonly plural: string;
  readonly titlePlural: string;
  readonly fields: Map<string, IField>;
  readonly operations: Map<string, IOperation>;
  readonly relations: Set<string>;
  readonly identity: Set<string>;
  readonly required: Set<string>;
  readonly indexed: Set<string>;
}

export interface IndexEntry {
  name: string;
  fields: { [field: string]: number };
  options: { sparse: boolean; unique: boolean };
}

export interface IndexEntryOptions {
  sparse?: boolean;
  unique?: boolean;
}

export interface EntityBasePersistence {
  indexes: {
    [index: string]: IndexEntry;
  };
}

export interface UIView {
  listName?: string[];
  quickSearch?: string[];
  hidden?: string[];
  edit?: string[];
  show?: string[];
  list?: string[];
  embedded?: string[];
  enum?: boolean;
  dictionary?: boolean;
}

export interface EntityBaseMetaInfo<P extends EntityBasePersistence>
  extends ElementMetaInfo {
  titlePlural: string;
  name: {
    plural: string;
    singular: string;
  };
  persistence: P;
  UI: UIView;
  acl: {
    create: string[];
    readOne: string[];
    readMany: string[];
    update: string[];
    delete: string[];
  };
}

export interface EntityBaseInternal extends ModelBaseInternal {
  fields: Map<string, IField>;
  operations: Map<string, IOperation>;
  relations: Set<string>;
  identity: Set<string>;
  required: Set<string>;
  indexed: Set<string>;
}

export interface EntityBaseInput<
  T extends EntityBaseMetaInfo<P>,
  P extends EntityBasePersistence
> extends ModelBaseInput<T> {
  plural?: string;
  titlePlural?: string;
  operations?: AsHash<OperationInput> | NamedArray<OperationInput>;
  fields?: AsHash<FieldInput> | NamedArray<FieldInput>;
}

export interface EntityBaseOutput<
  T extends EntityBaseMetaInfo<P>,
  P extends EntityBasePersistence
> extends ModelBaseOutput<T> {
  plural?: string;
  titlePlural?: string;
  operations?: NamedArray<OperationInput>;
  fields: NamedArray<FieldInput>;
}

export const entityBaseDefaultMetaInfo = {
  name: {},
  persistence: {
    indexes: {},
  },
  UI: {
    listName: [],
    quickSearch: [],
    hidden: [],
    edit: [],
    show: [],
    list: [],
    embedded: [],
  },
  acl: {
    create: [],
    read: [],
    update: [],
    delete: [],
  },
};
export const entityBaseDefaultInput = { metadata: entityBaseDefaultMetaInfo };

export class EntityBase<
  M extends EntityBaseMetaInfo<MP>,
  I extends EntityBaseInput<M, MP>,
  P extends EntityBaseInternal,
  MP extends EntityBasePersistence,
  O extends EntityBaseOutput<M, MP>
> extends ModelBase<M, I, P, O> implements IEntityBase<M, MP, I, O> {
  public get modelType(): MetaModelType {
    return 'entity-base';
  }

  constructor(init: I) {
    super(merge({}, entityBaseDefaultInput, init));
  }

  get plural(): string {
    return this.metadata.name.plural;
  }

  get titlePlural(): string {
    return this.metadata.titlePlural;
  }

  get relations(): Set<string> {
    return this[Internal].relations;
  }

  get required(): Set<string> {
    return this[Internal].required;
  }

  get identity(): Set<string> {
    return this[Internal].identity;
  }

  get fields(): Map<string, IField> {
    return this[Internal].fields;
  }

  get operations(): Map<string, IOperation> {
    return this[Internal].operations;
  }

  get indexed(): Set<string> {
    return this[Internal].indexed;
  }

  protected updateIndex(f: IField, options: IndexEntryOptions) {
    let indexName: string | string[];
    const key: 'indexed' | 'identity' = options.unique ? 'identity' : 'indexed';

    if (typeof f[key] === 'boolean') {
      indexName = f.name;
    } else if (Array.isArray(f[key])) {
      indexName = f[key] as string[];
    } else {
      indexName = (f[key] as string).split(' ');
      indexName = indexName.length > 1 ? indexName : indexName[0];
    }
    let entry = {
      fields: {
        [f.name]: 1,
      },
      options,
    };
    if (typeof indexName === 'string') {
      this.mergeIndex(indexName, { name: indexName, ...entry });
    } else {
      indexName.forEach(index => {
        this.mergeIndex(index, { name: index, ...entry });
      });
    }
  }

  protected mergeIndex(index: string, entry: any) {
    const indexes = this.metadata.persistence.indexes;
    if (this.metadata.persistence.indexes.hasOwnProperty(index)) {
      indexes[index] = merge(indexes[index], entry);
      if (Array.isArray(indexes[index].name)) {
        indexes[index].name = indexes[index].name[0];
      } else {
        indexes[index].name = indexes[index].name;
      }
    } else {
      indexes[index] = entry;
    }
  }

  public updateWith(input: Nullable<I>) {
    super.updateWith(input);

    assignValue<P, I, string>({
      src: this[Internal],
      input,
      field: 'name',
      effect: (src, value) => {
        src.name = capitalize(value.trim());
        this.metadata.name.singular = src.name;
      },
      required: true,
    });

    assignValue<M, I, string>({
      src: this.metadata,
      input,
      field: 'name.plural',
      inputField: 'plural',
      effect: (src, value) => {
        src.name.plural = capitalize(value.trim());
        if (src.name.plural === this.name) {
          src.name.plural = `All${this.name}`;
        }
      },
      setDefault: src => {
        src.name.plural = inflected.pluralize(this.name);
        if (src.name.plural === this.name) {
          src.name.plural = `All${this.name}`;
        }
      },
    });

    assignValue<M, I, string>({
      src: this.metadata,
      input,
      field: 'titlePlural',
      effect: (src, value) => (src.titlePlural = value.trim()),
      setDefault: src => (src.titlePlural = this.plural),
    });

    assignValue<P, I, AsHash<FieldInput> | NamedArray<FieldInput>>({
      src: this[Internal],
      input,
      field: 'fields',
      effect: (src, value) => {
        const relations = new Set();
        const identity = new Set();
        const required = new Set();
        const indexed = new Set();

        const fields = new Map(
          (Array.isArray(value) ? value : HashToArray(value))
            .map((fld, order) => {
              let field: Field;
              if (isSimpleInput(fld)) {
                field = new SimpleField(
                  merge({}, fld, {
                    order,
                    entity: this.name,
                  } as Partial<SimpleFieldInput>),
                );
                if (field.identity) {
                  identity.add(field.name);
                  this.updateIndex(field, { unique: true, sparse: true });
                }

                if (field.required) {
                  required.add(field.name);
                }
                if (field.indexed) {
                  indexed.add(field.name);
                  this.updateIndex(field, { sparse: true });
                }
              } else if (isRelationFieldInput(fld)) {
                field = new RelationField(
                  merge({}, fld, {
                    order,
                    entity: this.name,
                  } as Partial<RelationFieldInput>),
                );
                if (field.relation) {
                  relations.add(field.name);
                }
              } else {
                field = new EntityField(
                  merge({}, fld, {
                    order,
                    entity: this.name,
                  } as Partial<EntityFieldInput>),
                );
                if (field.relation) {
                  relations.add(field.name);
                }
              }
              return field;
            })
            .map(f => [f.name, f] as [string, Field]),
        );
        let f: SimpleField;
        if (fields.has('id')) {
          f = fields.get('id') as SimpleField;
          f.updateWith({ identity: true } as Nullable<SimpleFieldInput>);
          this.updateIndex(f, { unique: true });
        } else if (fields.has('_id')) {
          f = fields.get('_id') as SimpleField;
          f.updateWith({ identity: true } as Nullable<SimpleFieldInput>);
          this.updateIndex(f, { unique: true });
        } else {
          f = new SimpleField({
            ...DEFAULT_ID_FIELD,
            entity: this.name,
            order: -1,
          });
          fields.set(f.name, f);
          this.updateIndex(f, { unique: true });
        }

        f.makeIdentity();
        indexed.add(f.name);
        identity.add(f.name);
        required.add(f.name);

        src.relations = relations;
        src.identity = identity;
        src.required = required;
        src.indexed = indexed;
        src.fields = fields;
      },
      setDefault: src =>
        (src.fields = new Map([
          [
            DEFAULT_ID_FIELD.name,
            new SimpleField({
              ...DEFAULT_ID_FIELD,
              entity: this.name,
              order: -1,
            }),
          ],
        ] as [string, SimpleField][])),
    });

    assignValue<P, I, AsHash<OperationInput> | NamedArray<OperationInput>>({
      src: this[Internal],
      input,
      field: 'operations',
      effect: (src, value) => {
        src.operations = new Map(
          (Array.isArray(value) ? value : HashToArray(value))
            .map(
              (fld, order) =>
                new Operation(
                  merge({}, fld, {
                    entity: this.name,
                    order,
                  } as Partial<OperationInput>),
                ),
            )
            .map(f => [f.name, f] as [string, Operation]),
        );
      },
      setDefault: src => (src.operations = new Map<string, Operation>()),
    });
  }

  public ensureIndexes() {
    this.fields.forEach(f => {
      if (f.identity) {
        this.updateIndex(f, {});
      } else if (f.indexed) {
        this.updateIndex(f, {});
      }
    });
  }

  public toObject(): O {
    return merge({}, super.toObject(), {
      fields: MapToArray(this.fields, (name, value) => ({
        ...value.toObject(),
        name,
      })),
      operations: MapToArray(this.operations, (name, value) => ({
        ...value.toObject(),
        name,
      })),
    } as Partial<O>);
  }
}
