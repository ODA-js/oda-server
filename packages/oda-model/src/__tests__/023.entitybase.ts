import 'jest';
import { EntityBase, entityBaseDefaultMetaInfo } from '../entitybase';
import { MetaData } from '../element';
import { Field } from '../field';

describe('EntityBase', () => {
  it('default', () => {
    const res = new EntityBase({ name: 'A' });
    expect(res[MetaData]).toMatchObject(entityBaseDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('default has no Singular', () => {
    const res = new EntityBase({ name: 'Species' });
    expect(res.plural).toBe('AllSpecies');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('should Capitalize name', () => {
    const res = new EntityBase({ name: 'entity' });
    expect(res.name).toBe('Entity');
  });
  it('take TitlePlural from Plural name', () => {
    const res = new EntityBase({ name: 'Child', plural: 'Children  ' });
    expect(res.plural).toBe('Children');
    expect(res.titlePlural).toBe('Children');
  });
  it('should take Plural Title plural', () => {
    const res = new EntityBase({
      name: 'Species',
      plural: 'Species',
      titlePlural: 'All Species',
    });
    expect(res.titlePlural).toBe('All Species');
    expect(res.plural).toBe('AllSpecies');
  });
  it('should take operations', () => {
    const res = new EntityBase({
      name: 'A',
      operations: [
        {
          name: 'updateUser',
          entity: 'User',
          actionType: 'create',
          payload: [{ name: 'result', multiplicity: 'many' }],
          args: [{ name: 'id' }, { name: 'fields', multiplicity: 'many' }],
        },
      ],
    });
    expect(res.fields.size).toBe(1);
    expect(res.fields.get('id')).not.toBeUndefined();

    expect(res.operations.size).toBe(1);
    expect(res.toObject()).toMatchSnapshot('toObject with operations');
  });

  it('creates specified field with specified types', () => {
    const res = new EntityBase({
      name: 'A',
      fields: [
        {
          name: 'A',
        },
        {
          name: 'B',
          type: { type: 'enum', name: 'BBB' },
        },
        {
          name: 'C',
          type: { type: 'entity', name: 'CCC' },
        },
        {
          name: 'D',
          relation: {
            hasOne: 'CCC#',
          },
        },
      ],
    });
    expect(res.fields.size).toBe(5);
    expect(res.fields.get('id')).not.toBeUndefined();
    expect(res.fields.get('a')).not.toBeUndefined();
    expect(res.fields.get('b')).not.toBeUndefined();
    expect(res.fields.get('c')).not.toBeUndefined();
    expect(res.fields.get('d')).not.toBeUndefined();
    expect(res).toMatchSnapshot('object with fields');
    expect(res.toObject()).toMatchSnapshot('toObject with fields');
  });

  it('should accept non default identity field "id"', () => {
    const res = new EntityBase({ name: 'A', fields: [{ name: 'id' }] });
    expect(res.fields.size).toBe(1);
    expect(res.fields.get('id')).not.toBeUndefined();
    expect((res.fields.get('id') as Field).toObject()).toMatchSnapshot(
      'non default id fields',
    );
  });
  it('should accept non default identity field "_id"', () => {
    const res = new EntityBase({ name: 'A', fields: [{ name: '_id' }] });
    expect(res.fields.size).toBe(1);
    expect(res.fields.get('_id')).not.toBeUndefined();
    expect((res.fields.get('_id') as Field).toObject()).toMatchSnapshot(
      'non default _id fields',
    );
  });
});
