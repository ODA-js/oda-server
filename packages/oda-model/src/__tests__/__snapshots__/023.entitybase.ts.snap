// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityBase creates specified field with specified types: object with fields 1`] = `
EntityBase {
  Symbol(internal): Object {
    "fields": Map {
      "a" => SimpleField {
        Symbol(internal): Object {
          "args": Map {},
          "list": false,
          "name": "a",
          "type": "string",
        },
        Symbol(metadata): Object {
          "acl": Object {
            "read": Array [],
            "update": Array [],
          },
          "entity": "A",
          "order": 0,
          "persistence": Object {
            "derived": false,
            "identity": false,
            "indexed": false,
            "persistent": true,
            "required": false,
          },
        },
      },
      "b" => SimpleField {
        Symbol(internal): Object {
          "args": Map {},
          "list": false,
          "name": "b",
          "type": Object {
            "multiplicity": "one",
            "name": "BBB",
            "type": "enum",
          },
        },
        Symbol(metadata): Object {
          "acl": Object {
            "read": Array [],
            "update": Array [],
          },
          "entity": "A",
          "order": 1,
          "persistence": Object {
            "derived": false,
            "identity": false,
            "indexed": false,
            "persistent": true,
            "required": false,
          },
        },
      },
      "c" => EntityField {
        Symbol(internal): Object {
          "args": Map {},
          "list": false,
          "name": "c",
          "relation": HasOne {
            Symbol(internal): Object {
              "entity": "A",
              "field": "c",
              "hasOne": EntityReference {
                Symbol(internal): Object {
                  "backField": "id",
                  "entity": "CCC",
                  "field": "id",
                },
                Symbol(metadata): Object {},
              },
              "name": "AHasOneC",
            },
            Symbol(metadata): Object {
              "name": Object {
                "full": "AHasOneC",
                "normal": "AC",
                "short": "C",
              },
              "persistence": Object {
                "embedded": true,
                "single": true,
                "stored": false,
              },
              "verb": "HasOne",
            },
          },
          "type": Object {
            "multiplicity": "one",
            "name": "CCC",
            "type": "entity",
          },
        },
        Symbol(metadata): Object {
          "acl": Object {
            "create": Array [],
            "delete": Array [],
            "read": Array [],
            "update": Array [],
          },
          "entity": "A",
          "order": 2,
          "persistence": Object {
            "derived": false,
            "identity": false,
            "indexed": false,
            "persistent": false,
            "required": false,
          },
        },
      },
      "d" => RelationField {
        Symbol(internal): Object {
          "args": Map {},
          "name": "d",
          "relation": HasOne {
            Symbol(internal): Object {
              "entity": "A",
              "field": "d",
              "hasOne": EntityReference {
                Symbol(internal): Object {
                  "backField": "id",
                  "entity": "CCC",
                  "field": "id",
                },
                Symbol(metadata): Object {},
              },
              "name": "AHasOneD",
            },
            Symbol(metadata): Object {
              "name": Object {
                "full": "AHasOneD",
                "normal": "AD",
                "short": "D",
              },
              "persistence": Object {
                "embedded": false,
                "single": true,
                "stored": false,
              },
              "verb": "HasOne",
            },
          },
        },
        Symbol(metadata): Object {
          "acl": Object {
            "create": Array [],
            "delete": Array [],
            "read": Array [],
            "update": Array [],
          },
          "entity": "A",
          "order": 3,
          "persistence": Object {
            "derived": false,
            "identity": false,
            "indexed": false,
            "persistent": false,
            "required": false,
          },
        },
      },
      "id" => SimpleField {
        Symbol(internal): Object {
          "args": Map {},
          "idKey": EntityReference {
            Symbol(internal): Object {
              "backField": "id",
              "entity": "A",
              "field": "id",
            },
            Symbol(metadata): Object {},
          },
          "list": false,
          "name": "id",
          "type": "ID",
        },
        Symbol(metadata): Object {
          "acl": Object {
            "read": Array [],
            "update": Array [],
          },
          "entity": "A",
          "persistence": Object {
            "derived": false,
            "identity": true,
            "indexed": true,
            "persistent": true,
            "required": true,
          },
        },
      },
    },
    "identity": Set {
      "id",
    },
    "indexed": Set {
      "id",
    },
    "name": "A",
    "operations": Map {},
    "relations": Set {
      "c",
      "d",
    },
    "required": Set {
      "id",
    },
  },
  Symbol(metadata): Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "persistence": Object {
      "indexes": Object {
        "id": Object {
          "fields": Object {
            "id": 1,
          },
          "name": "id",
          "options": Object {
            "unique": true,
          },
        },
      },
    },
    "titlePlural": "As",
  },
}
`;

exports[`EntityBase creates specified field with specified types: toObject with fields 1`] = `
Object {
  "description": undefined,
  "fields": Array [
    Object {
      "args": Object {},
      "defaultValue": undefined,
      "derived": false,
      "description": undefined,
      "entity": "A",
      "identity": false,
      "indexed": false,
      "inheritedFrom": undefined,
      "list": false,
      "metadata": Object {
        "acl": Object {
          "read": Array [],
          "update": Array [],
        },
        "entity": "A",
        "order": 0,
        "persistence": Object {
          "derived": false,
          "identity": false,
          "indexed": false,
          "persistent": true,
          "required": false,
        },
      },
      "name": "a",
      "order": 0,
      "persistent": true,
      "required": false,
      "title": undefined,
      "type": "string",
    },
    Object {
      "args": Object {},
      "defaultValue": undefined,
      "derived": false,
      "description": undefined,
      "entity": "A",
      "identity": false,
      "indexed": false,
      "inheritedFrom": undefined,
      "list": false,
      "metadata": Object {
        "acl": Object {
          "read": Array [],
          "update": Array [],
        },
        "entity": "A",
        "order": 1,
        "persistence": Object {
          "derived": false,
          "identity": false,
          "indexed": false,
          "persistent": true,
          "required": false,
        },
      },
      "name": "b",
      "order": 1,
      "persistent": true,
      "required": false,
      "title": undefined,
      "type": Object {
        "multiplicity": "one",
        "name": "BBB",
        "type": "enum",
      },
    },
    Object {
      "args": Object {},
      "derived": false,
      "description": undefined,
      "entity": "A",
      "identity": false,
      "indexed": false,
      "inheritedFrom": undefined,
      "list": false,
      "metadata": Object {
        "acl": Object {
          "create": Array [],
          "delete": Array [],
          "read": Array [],
          "update": Array [],
        },
        "entity": "A",
        "order": 2,
        "persistence": Object {
          "derived": false,
          "identity": false,
          "indexed": false,
          "persistent": false,
          "required": false,
        },
      },
      "name": "c",
      "order": 2,
      "persistent": false,
      "required": false,
      "title": undefined,
      "type": Object {
        "multiplicity": "one",
        "name": "CCC",
        "type": "entity",
      },
    },
    Object {
      "args": Object {},
      "derived": false,
      "description": undefined,
      "entity": "A",
      "identity": false,
      "indexed": false,
      "inheritedFrom": undefined,
      "metadata": Object {
        "acl": Object {
          "create": Array [],
          "delete": Array [],
          "read": Array [],
          "update": Array [],
        },
        "entity": "A",
        "order": 3,
        "persistence": Object {
          "derived": false,
          "identity": false,
          "indexed": false,
          "persistent": false,
          "required": false,
        },
      },
      "name": "d",
      "order": 3,
      "persistent": false,
      "relation": Object {
        "embedded": false,
        "entity": "A",
        "field": "d",
        "hasOne": "id@CCC#id",
        "metadata": Object {
          "name": Object {
            "full": "AHasOneD",
            "normal": "AD",
            "short": "D",
          },
          "persistence": Object {
            "embedded": false,
            "single": true,
            "stored": false,
          },
          "verb": "HasOne",
        },
        "name": "AHasOneD",
        "opposite": undefined,
      },
      "required": false,
      "title": undefined,
    },
    Object {
      "args": Object {},
      "defaultValue": undefined,
      "derived": false,
      "description": undefined,
      "entity": "A",
      "identity": true,
      "indexed": true,
      "inheritedFrom": undefined,
      "list": false,
      "metadata": Object {
        "acl": Object {
          "read": Array [],
          "update": Array [],
        },
        "entity": "A",
        "persistence": Object {
          "derived": false,
          "identity": true,
          "indexed": true,
          "persistent": true,
          "required": true,
        },
      },
      "name": "id",
      "order": undefined,
      "persistent": true,
      "required": true,
      "title": undefined,
      "type": "ID",
    },
  ],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "persistence": Object {
      "indexes": Object {
        "id": Object {
          "fields": Object {
            "id": 1,
          },
          "name": "id",
          "options": Object {
            "unique": true,
          },
        },
      },
    },
    "titlePlural": "As",
  },
  "name": "A",
  "operations": Array [],
  "title": undefined,
}
`;

exports[`EntityBase default has no Singular: toObject 1`] = `
Object {
  "description": undefined,
  "fields": Array [
    Object {
      "args": Object {},
      "defaultValue": undefined,
      "derived": false,
      "description": undefined,
      "entity": "Species",
      "identity": true,
      "indexed": true,
      "inheritedFrom": undefined,
      "list": false,
      "metadata": Object {
        "acl": Object {
          "read": Array [],
          "update": Array [],
        },
        "entity": "Species",
        "persistence": Object {
          "derived": false,
          "identity": true,
          "indexed": true,
          "persistent": true,
          "required": true,
        },
      },
      "name": "id",
      "order": undefined,
      "persistent": true,
      "required": true,
      "title": undefined,
      "type": "ID",
    },
  ],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "name": Object {
      "plural": "AllSpecies",
      "singular": "Species",
    },
    "persistence": Object {
      "indexes": Object {},
    },
    "titlePlural": "AllSpecies",
  },
  "name": "Species",
  "operations": Array [],
  "title": undefined,
}
`;

exports[`EntityBase default: toObject 1`] = `
Object {
  "description": undefined,
  "fields": Array [
    Object {
      "args": Object {},
      "defaultValue": undefined,
      "derived": false,
      "description": undefined,
      "entity": "A",
      "identity": true,
      "indexed": true,
      "inheritedFrom": undefined,
      "list": false,
      "metadata": Object {
        "acl": Object {
          "read": Array [],
          "update": Array [],
        },
        "entity": "A",
        "persistence": Object {
          "derived": false,
          "identity": true,
          "indexed": true,
          "persistent": true,
          "required": true,
        },
      },
      "name": "id",
      "order": undefined,
      "persistent": true,
      "required": true,
      "title": undefined,
      "type": "ID",
    },
  ],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "persistence": Object {
      "indexes": Object {},
    },
    "titlePlural": "As",
  },
  "name": "A",
  "operations": Array [],
  "title": undefined,
}
`;

exports[`EntityBase should accept non default identity field: non default id fields 1`] = `
Object {
  "args": Object {},
  "defaultValue": undefined,
  "derived": false,
  "description": undefined,
  "entity": "A",
  "identity": true,
  "indexed": true,
  "inheritedFrom": undefined,
  "list": false,
  "metadata": Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "entity": "A",
    "order": 0,
    "persistence": Object {
      "derived": false,
      "identity": true,
      "indexed": true,
      "persistent": true,
      "required": true,
    },
  },
  "name": "id",
  "order": 0,
  "persistent": true,
  "required": true,
  "title": undefined,
  "type": "string",
}
`;

exports[`EntityBase should take operations: toObject with operations 1`] = `
Object {
  "description": undefined,
  "fields": Array [
    Object {
      "args": Object {},
      "defaultValue": undefined,
      "derived": false,
      "description": undefined,
      "entity": "A",
      "identity": true,
      "indexed": true,
      "inheritedFrom": undefined,
      "list": false,
      "metadata": Object {
        "acl": Object {
          "read": Array [],
          "update": Array [],
        },
        "entity": "A",
        "persistence": Object {
          "derived": false,
          "identity": true,
          "indexed": true,
          "persistent": true,
          "required": true,
        },
      },
      "name": "id",
      "order": undefined,
      "persistent": true,
      "required": true,
      "title": undefined,
      "type": "ID",
    },
  ],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "persistence": Object {
      "indexes": Object {},
    },
    "titlePlural": "As",
  },
  "name": "A",
  "operations": Array [
    Object {
      "actionType": "create",
      "args": Array [
        Object {
          "defaultValue": undefined,
          "metadata": Object {},
          "multiplicity": "one",
          "name": "id",
          "required": false,
          "type": "string",
        },
        Object {
          "defaultValue": undefined,
          "metadata": Object {},
          "multiplicity": "many",
          "name": "fields",
          "required": false,
          "type": "string",
        },
      ],
      "description": undefined,
      "inheritedFrom": undefined,
      "metadata": Object {
        "acl": Object {
          "execute": Array [],
        },
      },
      "name": "updateUser",
      "payload": Array [
        Object {
          "defaultValue": undefined,
          "metadata": Object {},
          "multiplicity": "many",
          "name": "result",
          "required": false,
          "type": "string",
        },
      ],
      "title": undefined,
    },
  ],
  "title": undefined,
}
`;
